Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -0,0 +1,314 @@
+ * `WeierstrassCurve.Jacobian.toAffine`: the Weierstrass curve in affine coordinates.
+ * `WeierstrassCurve.Jacobian.nonsingular`: the nonsingular condition on a point representative.
+ * `WeierstrassCurve.Jacobian.nonsingular_lift`: the nonsingular condition on a point class.
+
+## Implementation notes
+
+A point representative is implemented as a term `P` of type `Fin 3 → R`, which allows for the vector
+notation `![x, y, z]`. However, `P` is not definitionally equivalent to the expanded vector
+`![P x, P y, P z]`, so the auxiliary lemma `fin3_def` can be used to convert between the two forms.
+The equivalence of two point representatives `P` and `Q` is implemented as an equivalence of orbits
+of the action of `Rˣ`, or equivalently that there is some unit `u` of `R` such that `P = u • Q`.
+However, `u • Q` is again not definitionally equal to `![u² * Q x, u³ * Q y, u * Q z]`, so the
+auxiliary lemmas `smul_fin3` and `smul_fin3_ext` can be used to convert between the two forms.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, rational point, Jacobian coordinates
+-/
+
+local notation "x" => 0
+
+local notation "y" => 1
+
+local notation "z" => 2
+
+local macro "matrix_simp" : tactic =>
+  `(tactic| simp only [Matrix.head_cons, Matrix.tail_cons, Matrix.smul_empty, Matrix.smul_cons,
+              Matrix.cons_val_zero, Matrix.cons_val_one, Matrix.cons_val_two])
+
+universe u
+
+lemma fin3_def {R : Type u} (P : Fin 3 → R) : P = ![P x, P y, P z] := by
+  ext n; fin_cases n <;> rfl
+
+private instance {R : Type u} [CommRing R] : SMul Rˣ <| Fin 3 → R :=
+  ⟨fun u P => ![u ^ 2 * P x, u ^ 3 * P y, u * P z]⟩
+
+lemma smul_fin3 {R : Type u} [CommRing R] (P : Fin 3 → R) (u : Rˣ) :
+    u • P = ![u ^ 2 * P x, u ^ 3 * P y, u * P z] :=
+  rfl
+
+lemma smul_fin3_ext {R : Type u} [CommRing R] (P : Fin 3 → R) (u : Rˣ) :
+    (u • P) 0 = u ^ 2 * P x ∧ (u • P) 1 = u ^ 3 * P y ∧ (u • P) 2 = u * P z :=
+  ⟨rfl, rfl, rfl⟩
+
+private instance {R : Type u} [CommRing R] : MulAction Rˣ <| Fin 3 → R where
+  one_smul := fun _ => by
+    simp only [smul_fin3, Units.val_one, one_pow, one_mul, ← fin3_def]
+  mul_smul := fun u v P => by
+    simp only [smul_fin3, Units.val_mul, mul_pow, mul_assoc]
+    matrix_simp
+
+/-! ## Weierstrass curves -/
+
+/-- An abbreviation for a Weierstrass curve in Jacobian coordinates. -/
+abbrev WeierstrassCurve.Jacobian :=
+  WeierstrassCurve
+
+namespace WeierstrassCurve.Jacobian
+
+open MvPolynomial
+
+local macro "eval_simp" : tactic =>
+  `(tactic| simp only [eval_C, eval_X, eval_add, eval_sub, eval_mul, eval_pow])
+
+variable (R : Type u) [CommRing R]
+
+/-- The equivalence setoid for a point representative. -/
+def PointSetoid : Setoid <| Fin 3 → R :=
+  MulAction.orbitRel Rˣ <| Fin 3 → R
+
+attribute [local instance] PointSetoid
+
+/-- The equivalence class of a point representative. -/
+abbrev PointClass : Type u :=
```

Comment:
The same question about `Projectivization` as in the PR about projective coords applies here as well.

Reply:
This will be discussed in the projective coordinates PR.